libfeed/feed-model-ordering.c

```diff
@@ -0,0 +1,401 @@
+/* Copyright 2018 Endless Mobile, Inc.
+ *
+ * eos-discovery-feed is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 2.1 of the
+ * License, or (at your option) any later version.
+ *
+ * eos-discovery-feed is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with eos-discovery-feed.  If not, see
+ * <http://www.gnu.org/licenses/>.
+ */
+
+#include <glib.h>
+
+#include "feed-base-card-store.h"
+#include "feed-model-ordering.h"
+#include "feed-orderable-model.h"
+
+static gboolean
+models_remaining (GHashTable *model_map,
+                  GHashTable *app_indices_for_card_type)
+{
+  GHashTableIter iter;
+  EosDiscoveryFeedCardStoreType app_indices_for_card_type_key;
+  guint app_indices_for_card_type_value;
+
+  g_hash_table_iter_init (&iter, app_indices_for_card_type);
+  while (g_hash_table_iter_next (&iter,
+                                 (gpointer) &app_indices_for_card_type_key,
+                                 (gpointer) &app_indices_for_card_type_value))
+    {
+      GPtrArray *apps_for_card_type = g_hash_table_lookup (model_map,
+                                                           GINT_TO_POINTER (app_indices_for_card_type_key));
+
+      if (apps_for_card_type != NULL &&
+          app_indices_for_card_type_value < apps_for_card_type->len)
+        return TRUE;
+    }
+
+  return FALSE;
+}
+
+static void
+insert_model_into_map (GHashTable                     *map,
+                       EosDiscoveryFeedCardStoreType   type,
+                       const gchar                    *source,
+                       EosDiscoveryFeedOrderableModel *orderable_model)
+{
+  GHashTable *app_map = g_hash_table_lookup (map, GINT_TO_POINTER (type));
+  GPtrArray *model_list = NULL;
+  const gchar *resolved_source = source != NULL ? source : "no-app";
+
+  if (app_map == NULL)
+    {
+      app_map = g_hash_table_new_full (g_str_hash,
+                                       g_str_equal,
+                                       g_free,
+                                       (GDestroyNotify) g_ptr_array_unref);
+      g_hash_table_insert (map, GINT_TO_POINTER (type), app_map);
+    }
+
+  model_list = g_hash_table_lookup (app_map, resolved_source);
+
+  if (model_list == NULL)
+    {
+      model_list = g_ptr_array_new_with_free_func (g_object_unref);
+      g_hash_table_insert (app_map, g_strdup (resolved_source), model_list);
+    }
+
+  g_ptr_array_add (model_list, g_object_ref (orderable_model));
+}
+
+typedef struct _IndexableDescriptorEntry {
+  const gchar *app_id;
```

Comment:
It seems to me that this field was never used in the original code. I'll remove the enclosing struct here.